src/plugins/ciu/ciuPlugin.js

```diff
@@ -0,0 +1,118 @@
+const fs = require('fs');
+const { messageToFactoid } = require('../factoids/factoidsPlugin');
+
+const AGENTS_TO_SHOW = {
+  ie: 'IE',
+  edge: 'Edge',
+  firefox: 'FF',
+  chrome: 'Chrome',
+  opera: 'Opera',
+  safari: 'Safari',
+  ios_saf: 'iOS',
+  android: 'Android',
+};
+
+function getVersions(agents, feature) {
+  return Object.entries(AGENTS_TO_SHOW)
+    .map(([agent, agentName]) => {
+      // agents versions strings are sorted
+      const versions = agents[agent].versions.filter(Boolean);
+      const versionFull = versions.find(
+        (v) => v && !/[nupac]/g.test(feature.stats[agent][v]),
+      );
+
+      if (versionFull) return `${agentName} ${versionFull}`;
+
+      const versionPartial = versions.find(
+        (v) => v && !/[nu]/g.test(feature.stats[agent][v]),
+      );
+
+      if (versionPartial) return `${agentName} ${versionPartial}~`;
+
+      return null;
+    })
+    .filter(Boolean);
+}
+
+// return all permutations of xs items, including sub-permutations
+// e.g. perms([0,1]) → [[0,1], [1,0], [1], [0]]
+function perms(xs) {
+  if (xs.length === 1) return [xs];
+  return xs.flatMap((x, i) =>
+    perms([...xs.slice(0, i), ...xs.slice(i + 1)]).flatMap((vs) => [vs, [x, ...vs]]),
```

Comment:
```suggestion
    perms(xs.slice(0, i).concat(xs.slice(i + 1))).flatMap((vs) => [vs, [x, ...vs]]),
```
array spread is slow and often unnecessary

Reply:
if you want, but I thought performance shouldn't matter :p

Reply:
lol touché :-p but it's also messier imo, i like the concat form better